src/main/java/net/minecraftforge/fluids/FluidTank.java

```diff
@@ -175,11 +198,38 @@ public FluidStack drain(int maxDrain, boolean doDrain)
                 fluid = null;
             }
 
+            onContentsChanged();
+
             if (tile != null)
             {
                 FluidEvent.fireEvent(new FluidEvent.FluidDrainingEvent(fluid, tile.getWorld(), tile.getPos(), this, drained));
             }
         }
         return stack;
     }
+
+    /**
+     * Returns true if the tank can be filled with this type of fluid.
+     * Used as a filter for fluid types.
+     * Does not consider the current state of the tank, only whether it could ever fill with this type of fluid.
+     */
+    public boolean canFillFluidType(FluidStack fluid)
+    {
```

Comment:
The original definition was unclear, many people (myself included) interpreted it as a quicker check like a simulated drain or fill, and it just opens the door for buggy fluid handling.

I want the API to be clearly divided between:

- "state" which is like an immutable overview of contents of the tank. 
- "properties" which would be properties that are not dependent on the state of the tank, and include transaction abilities like `canFillFluidType`,
-  "transaction" fill/drain, which lets you interact without needing to know state or info.

The current TankInfo mainly gives the "state" of the tank and should be expanded on.